File: java/src/test/java/io/github/muthuishere/toolnexus/McpServeTest.java
package io.github.muthuishere.toolnexus;

import io.modelcontextprotocol.client.McpClient;
import io.modelcontextprotocol.client.McpSyncClient;
import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
import io.modelcontextprotocol.spec.McpSchema;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * MCP inbound — "serve as an MCP server" suite (SPEC §7C). Mirrors the nine
 * {@code "mcp inbound:"} tests in the JS reference ({@code js/test/unit.test.ts}).
 *
 * <p>The Java MCP SDK ships no in-memory/linked transport (unlike the JS SDK's
 * {@code InMemoryTransport.createLinkedPair()}), so — as SPEC §7C permits — these
 * tests connect the port's own MCP <b>client</b> to the served toolkit over an
 * ephemeral streamable-HTTP port and assert the {@code tools/list} /
 * {@code tools/call} round-trips. Hermetic: localhost only, no real LLM.
 */
class McpServeTest {

    private static final HttpClient HTTP = HttpClient.newHttpClient();

    // Schema shared by the test tools: {type, properties:{text:string}, required:[text], additionalProperties:false}.
    private static Map<String, Object> textSchema() {
        Map<String, Object> textProp = new LinkedHashMap<>();
        textProp.put("type", "string");
        Map<String, Object> props = new LinkedHashMap<>();
        props.put("text", textProp);
        Map<String, Object> schema = new LinkedHashMap<>();
        schema.put("type", "object");
        schema.put("properties", props);
        schema.put("required", List.of("text"));
        schema.put("additionalProperties", false);
        return schema;
    }

    private static Tool echoTool() {
        return NativeTool.of("echo", "echoes back", textSchema(), args -> args.get("text"));
    }

    private static Tool boomTool() {
        return NativeTool.of("boom", "always throws", textSchema(), args -> {
            throw new RuntimeException("kaboom");
        });
    }

    /** Connect an MCP client to a served toolkit's {@code /mcp} endpoint and initialize it. */
    private static McpSyncClient connect(String baseUrl) {
        HttpClientStreamableHttpTransport transport = HttpClientStreamableHttpTransport
                .builder(baseUrl + "/mcp")
                .jsonMapper(McpServe.JSON_MAPPER)
                .build();
        McpSyncClient client = McpClient.sync(transport)
                .clientInfo(new McpSchema.Implementation("test-client", "1.0.0"))
                .build();
        client.initialize();
        return client;
    }

    private static String firstText(McpSchema.CallToolResult res) {
        for (McpSchema.Content c : res.content()) {
            if (c instanceof McpSchema.TextContent t) return t.text();
        }
        return null;
    }

    // -----------------------------------------------------------------------
    // 1. exposedMcpTools filters by name, omit ⇒ all.
    // -----------------------------------------------------------------------

    @Test
    void exposedMcpToolsFiltersByNameOmitAll() {
        List<Tool> tools = List.of(echoTool(), boomTool());
        assertEquals(List.of("echo", "boom"), names(McpServe.exposedMcpTools(tools, null)));
        assertEquals(List.of("echo"),
                names(McpServe.exposedMcpTools(tools, new McpServe.MCPServeConfig().tools(List.of("echo")))));
        // unknown names in the filter are ignored, not an error
        assertEquals(List.of("echo"),
                names(McpServe.exposedMcpTools(tools, new McpServe.MCPServeConfig().tools(List.of("echo", "nope")))));
    }

    private static List<String> names(List<Tool> tools) {
        List<String> out = new ArrayList<>();
        for (Tool t : tools) out.add(t.name());
        return out;
    }

    // -----------------------------------------------------------------------
    // 2. initialize reports serverInfo from the profile.
    // -----------------------------------------------------------------------

    @Test
    void initializeReportsServerInfoFromProfile() {
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig().name("gateway").version("2.0.0")));
            McpSyncClient client = connect(srv.url());
            try {
                McpSchema.Implementation info = client.getServerInfo();
                assertEquals("gateway", info.name());
                assertEquals("2.0.0", info.version());
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 3. tools/list advertises unified tools with inputSchema=parameters.
    // -----------------------------------------------------------------------

    @Test
    void toolsListAdvertisesUnifiedToolsWithInputSchema() {
        Tool echo = echoTool();
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echo, boomTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig()));
            McpSyncClient client = connect(srv.url());
            try {
                List<McpSchema.Tool> tools = client.listTools().tools();
                List<String> got = new ArrayList<>();
                for (McpSchema.Tool t : tools) got.add(t.name());
                got.sort(String::compareTo);
                assertEquals(List.of("boom", "echo"), got);
                McpSchema.Tool advertised = tools.stream()
                        .filter(t -> t.name().equals("echo")).findFirst().orElseThrow();
                assertEquals(echo.inputSchema(), advertised.inputSchema());
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 4. mcp.tools narrows the advertised surface.
    // -----------------------------------------------------------------------

    @Test
    void mcpToolsNarrowsTheAdvertisedSurface() {
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool(), boomTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig().tools(List.of("echo"))));
            McpSyncClient client = connect(srv.url());
            try {
                List<String> got = names(new ArrayList<>());
                for (McpSchema.Tool t : client.listTools().tools()) got.add(t.name());
                assertEquals(List.of("echo"), got);
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 5. tools/call dispatches to execute → text content, isError false, onCall fires.
    // -----------------------------------------------------------------------

    @Test
    void toolsCallDispatchesToExecuteTextContent() {
        List<McpServe.OnCallEvent> calls = new CopyOnWriteArrayList<>();
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig()).onCall(calls::add));
            McpSyncClient client = connect(srv.url());
            try {
                McpSchema.CallToolResult res = client.callTool(
                        new McpSchema.CallToolRequest("echo", Map.of("text", "hi")));
                assertFalse(Boolean.TRUE.equals(res.isError()));
                assertEquals("hi", firstText(res));
                assertEquals(1, calls.size());
                assertEquals("echo", calls.get(0).name());
                assertEquals("native", calls.get(0).source());
                assertFalse(calls.get(0).isError());
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 6. erroring tool → isError result, server survives.
    // -----------------------------------------------------------------------

    @Test
    void erroringToolBecomesIsErrorResultServerSurvives() {
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool(), boomTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig()));
            McpSyncClient client = connect(srv.url());
            try {
                McpSchema.CallToolResult bad = client.callTool(
                        new McpSchema.CallToolRequest("boom", Map.of()));
                assertTrue(Boolean.TRUE.equals(bad.isError()));
                assertTrue(firstText(bad).contains("kaboom"), "error text: " + firstText(bad));

                // server keeps serving
                McpSchema.CallToolResult ok = client.callTool(
                        new McpSchema.CallToolRequest("echo", Map.of("text", "still up")));
                assertFalse(Boolean.TRUE.equals(ok.isError()));
                assertEquals("still up", firstText(ok));
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 7. streamable-HTTP /mcp round-trip via serve().
    // -----------------------------------------------------------------------

    @Test
    void streamableHttpRoundTripViaServe() {
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().mcp(new McpServe.MCPServeConfig().name("http-gw")));
            McpSyncClient client = connect(srv.url());
            try {
                assertEquals("http-gw", client.getServerInfo().name());
                boolean hasEcho = client.listTools().tools().stream().anyMatch(t -> t.name().equals("echo"));
                assertTrue(hasEcho);
                McpSchema.CallToolResult res = client.callTool(
                        new McpSchema.CallToolRequest("echo", Map.of("text", "over http")));
                assertEquals("over http", firstText(res));
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 8. absent profile ⇒ no /mcp surface (A2A handler treats it as unknown method).
    // -----------------------------------------------------------------------

    @Test
    @SuppressWarnings("unchecked")
    void absentProfileNoMcpSurface() throws Exception {
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(echoTool()))) {
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0",
                    new Toolkit.ServeOptions().a2a(new A2AServer.A2AConfig().name("only-a2a")));
            try {
                Map<String, Object> rpc = new LinkedHashMap<>();
                rpc.put("jsonrpc", "2.0");
                rpc.put("id", 1);
                rpc.put("method", "tools/list");
                rpc.put("params", Map.of());
                HttpResponse<String> res = HTTP.send(
                        HttpRequest.newBuilder(URI.create(srv.url() + "/mcp"))
                                .header("Content-Type", "application/json")
                                .header("Accept", "application/json, text/event-stream")
                                .POST(HttpRequest.BodyPublishers.ofString(Json.stringify(rpc)))
                                .build(),
                        HttpResponse.BodyHandlers.ofString());
                Map<String, Object> body = Json.toMap(res.body());
                // A2A POST handler treats it as an unknown JSON-RPC method (not an MCP endpoint).
                assertTrue(body.containsKey("error"), "expected a JSON-RPC error, not an MCP result");
                assertNull(body.get("result"));
            } finally {
                srv.stop();
            }
        }
    }

    // -----------------------------------------------------------------------
    // 9. top-level mcpServer config block is picked up by serve().
    // -----------------------------------------------------------------------

    @Test
    void topLevelMcpServerConfigBlockIsPickedUp() {
        Map<String, Object> config = new LinkedHashMap<>();
        config.put("mcpServer", Map.of("name", "from-config"));
        try (Toolkit tk = Toolkit.create(new Toolkit.Options()
                .mcpConfig(config).builtins(false).extraTools(echoTool()))) {
            // serve() with no inline mcp falls back to the config block.
            A2AServer.ServeHandle srv = tk.serve("127.0.0.1:0", new Toolkit.ServeOptions());
            McpSyncClient client = connect(srv.url());
            try {
                assertEquals("from-config", client.getServerInfo().name());
            } finally {
                client.closeGracefully();
                srv.stop();
            }
        }
    }
}

File: golang/mcpserve_test.go
// MCP (inbound) tests — a toolkit exposed AS an MCP server (SPEC §7C). The
// in-process client (mark3labs NewInProcessClient) drives tools/list + tools/call
// against a built server; the streamable-HTTP path is exercised through a served
// toolkit with the mcp-go streamable HTTP client. Mirrors the js "mcp inbound:"
// tests. Run: go test -race .

package toolnexus

import (
	"context"
	"encoding/json"
	"errors"
	"reflect"
	"strings"
	"testing"

	mcpclient "github.com/mark3labs/mcp-go/client"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// echoTool echoes its `text` argument back verbatim (source: native).
var echoTool = Tool{
	Name:        "echo",
	Description: "echo back the text",
	Source:      SourceNative,
	InputSchema: JSONSchema{
		"type":                 "object",
		"properties":           map[string]any{"text": map[string]any{"type": "string"}},
		"required":             []any{"text"},
		"additionalProperties": false,
	},
	Execute: func(args map[string]any, _ *ToolContext) (ToolResult, error) {
		text, _ := args["text"].(string)
		return ToolResult{Output: text}, nil
	},
}

// boomTool always errors — proves an Execute error becomes an isError result
// without crashing the server.
var boomTool = Tool{
	Name:        "boom",
	Description: "always explodes",
	Source:      SourceNative,
	InputSchema: JSONSchema{"type": "object", "properties": map[string]any{}, "additionalProperties": false},
	Execute: func(_ map[string]any, _ *ToolContext) (ToolResult, error) {
		return ToolResult{}, errors.New("kaboom")
	},
}

// connectInProc builds an initialized in-process MCP client for the server.
func connectInProc(t *testing.T, srv *server.MCPServer) *mcpclient.Client {
	t.Helper()
	c, err := mcpclient.NewInProcessClient(srv)
	if err != nil {
		t.Fatalf("NewInProcessClient: %v", err)
	}
	if err := c.Start(context.Background()); err != nil {
		t.Fatalf("client Start: %v", err)
	}
	if _, err := c.Initialize(context.Background(), mcp.InitializeRequest{}); err != nil {
		t.Fatalf("client Initialize: %v", err)
	}
	return c
}

func toolNames(tools []mcp.Tool) []string {
	names := make([]string, len(tools))
	for i, t := range tools {
		names[i] = t.Name
	}
	return names
}

func TestMcpInboundExposedFilter(t *testing.T) {
	tools := []Tool{echoTool, boomTool}
	if got := names(ExposedMcpTools(tools, nil)); !reflect.DeepEqual(got, []string{"echo", "boom"}) {
		t.Fatalf("omit ⇒ all: got %v", got)
	}
	if got := names(ExposedMcpTools(tools, &MCPServeConfig{Tools: []string{"echo"}})); !reflect.DeepEqual(got, []string{"echo"}) {
		t.Fatalf("filter: got %v", got)
	}
	// unknown names in the filter are ignored, not an error
	if got := names(ExposedMcpTools(tools, &MCPServeConfig{Tools: []string{"echo", "nope"}})); !reflect.DeepEqual(got, []string{"echo"}) {
		t.Fatalf("unknown ignored: got %v", got)
	}
}

func names(tools []Tool) []string {
	out := make([]string, len(tools))
	for i, t := range tools {
		out[i] = t.Name
	}
	return out
}

func TestMcpInboundInitializeServerInfo(t *testing.T) {
	srv := buildMcpServer([]Tool{echoTool}, &MCPServeConfig{Name: "gateway", Version: "2.0.0"}, nil)
	c, err := mcpclient.NewInProcessClient(srv)
	if err != nil {
		t.Fatalf("NewInProcessClient: %v", err)
	}
	defer c.Close()
	if err := c.Start(context.Background()); err != nil {
		t.Fatalf("Start: %v", err)
	}
	res, err := c.Initialize(context.Background(), mcp.InitializeRequest{})
	if err != nil {
		t.Fatalf("Initialize: %v", err)
	}
	if res.ServerInfo.Name != "gateway" {
		t.Fatalf("serverInfo.name = %q, want gateway", res.ServerInfo.Name)
	}
	if res.ServerInfo.Version != "2.0.0" {
		t.Fatalf("serverInfo.version = %q, want 2.0.0", res.ServerInfo.Version)
	}
}

func TestMcpInboundToolsListSchema(t *testing.T) {
	srv := buildMcpServer([]Tool{echoTool, boomTool}, nil, nil)
	c := connectInProc(t, srv)
	defer c.Close()

	res, err := c.ListTools(context.Background(), mcp.ListToolsRequest{})
	if err != nil {
		t.Fatalf("ListTools: %v", err)
	}
	got := toolNames(res.Tools)
	if len(got) != 2 || !contains(got, "echo") || !contains(got, "boom") {
		t.Fatalf("tool names = %v, want echo+boom", got)
	}
	var echo *mcp.Tool
	for i := range res.Tools {
		if res.Tools[i].Name == "echo" {
			echo = &res.Tools[i]
		}
	}
	if echo == nil {
		t.Fatal("echo tool missing")
	}
	// inputSchema round-trips byte-equal to the tool's parameters.
	if !schemasEqual(t, echo.InputSchema, echoTool.InputSchema) {
		gotJSON, _ := json.Marshal(echo.InputSchema)
		wantJSON, _ := json.Marshal(echoTool.InputSchema)
		t.Fatalf("inputSchema = %s, want %s", gotJSON, wantJSON)
	}
}

func TestMcpInboundToolsFilterNarrows(t *testing.T) {
	cfg := &MCPServeConfig{Tools: []string{"echo"}}
	srv := buildMcpServer(ExposedMcpTools([]Tool{echoTool, boomTool}, cfg), cfg, nil)
	c := connectInProc(t, srv)
	defer c.Close()

	res, err := c.ListTools(context.Background(), mcp.ListToolsRequest{})
	if err != nil {
		t.Fatalf("ListTools: %v", err)
	}
	if got := toolNames(res.Tools); !reflect.DeepEqual(got, []string{"echo"}) {
		t.Fatalf("filtered list = %v, want [echo]", got)
	}
}

func TestMcpInboundCallDispatch(t *testing.T) {
	var calls []OnCallEvent
	srv := buildMcpServer([]Tool{echoTool}, nil, func(ev OnCallEvent) { calls = append(calls, ev) })
	c := connectInProc(t, srv)
	defer c.Close()

	req := mcp.CallToolRequest{}
	req.Params.Name = "echo"
	req.Params.Arguments = map[string]any{"text": "hi"}
	res, err := c.CallTool(context.Background(), req)
	if err != nil {
		t.Fatalf("CallTool: %v", err)
	}
	if res.IsError {
		t.Fatalf("isError = true, want false")
	}
	if text := textOf(t, res); text != "hi" {
		t.Fatalf("content text = %q, want hi", text)
	}
	if len(calls) != 1 {
		t.Fatalf("onCall count = %d, want 1", len(calls))
	}
	if calls[0].Name != "echo" || calls[0].Source != SourceNative || calls[0].IsError {
		t.Fatalf("onCall event = %+v", calls[0])
	}
}

func TestMcpInboundErrorSurvives(t *testing.T) {
	srv := buildMcpServer([]Tool{echoTool, boomTool}, nil, nil)
	c := connectInProc(t, srv)
	defer c.Close()

	boom := mcp.CallToolRequest{}
	boom.Params.Name = "boom"
	boom.Params.Arguments = map[string]any{}
	bad, err := c.CallTool(context.Background(), boom)
	if err != nil {
		t.Fatalf("CallTool(boom): %v", err)
	}
	if !bad.IsError {
		t.Fatalf("boom isError = false, want true")
	}
	if text := textOf(t, bad); !strings.Contains(text, "kaboom") {
		t.Fatalf("boom text = %q, want to contain kaboom", text)
	}

	// server keeps serving
	ok := mcp.CallToolRequest{}
	ok.Params.Name = "echo"
	ok.Params.Arguments = map[string]any{"text": "still up"}
	res, err := c.CallTool(context.Background(), ok)
	if err != nil {
		t.Fatalf("CallTool(echo) after boom: %v", err)
	}
	if res.IsError {
		t.Fatalf("echo isError = true after boom")
	}
	if text := textOf(t, res); text != "still up" {
		t.Fatalf("echo text = %q, want 'still up'", text)
	}
}

func TestMcpInboundStreamableHTTPRoundTrip(t *testing.T) {
	tk, err := CreateToolkit(context.Background(), Options{Builtins: false, ExtraTools: []Tool{echoTool}})
	if err != nil {
		t.Fatalf("CreateToolkit: %v", err)
	}
	defer tk.Close()

	handle, err := tk.Serve("127.0.0.1:0", ServeOptions{MCP: &MCPServeConfig{Name: "http-gw"}})
	if err != nil {
		t.Fatalf("Serve: %v", err)
	}
	defer handle.Stop()

	c, err := mcpclient.NewStreamableHttpClient(handle.URL + "/mcp")
	if err != nil {
		t.Fatalf("NewStreamableHttpClient: %v", err)
	}
	defer c.Close()
	if err := c.Start(context.Background()); err != nil {
		t.Fatalf("Start: %v", err)
	}
	initRes, err := c.Initialize(context.Background(), mcp.InitializeRequest{})
	if err != nil {
		t.Fatalf("Initialize: %v", err)
	}
	if initRes.ServerInfo.Name != "http-gw" {
		t.Fatalf("serverInfo.name = %q, want http-gw", initRes.ServerInfo.Name)
	}
	list, err := c.ListTools(context.Background(), mcp.ListToolsRequest{})
	if err != nil {
		t.Fatalf("ListTools: %v", err)
	}
	if !contains(toolNames(list.Tools), "echo") {
		t.Fatalf("tools = %v, want to contain echo", toolNames(list.Tools))
	}
	req := mcp.CallToolRequest{}
	req.Params.Name = "echo"
	req.Params.Arguments = map[string]any{"text": "over http"}
	res, err := c.CallTool(context.Background(), req)
	if err != nil {
		t.Fatalf("CallTool: %v", err)
	}
	if text := textOf(t, res); text != "over http" {
		t.Fatalf("text = %q, want 'over http'", text)
	}
}

func TestMcpInboundAbsentNoSurface(t *testing.T) {
	tk, err := CreateToolkit(context.Background(), Options{Builtins: false, ExtraTools: []Tool{echoTool}})
	if err != nil {
		t.Fatalf("CreateToolkit: %v", err)
	}
	defer tk.Close()

	client := CreateClient(ClientOptions{BaseURL: "http://unused", Style: StyleOpenAI, Model: "x", APIKey: "k"})
	handle, err := tk.Serve("127.0.0.1:0", ServeOptions{Client: client, A2A: &A2AConfig{Name: "only-a2a"}}) // no MCP
	if err != nil {
		t.Fatalf("Serve: %v", err)
	}
	defer handle.Stop()

	// A POST /mcp with an MCP-style JSON-RPC body is handled by the A2A POST
	// handler as an unknown method (not an MCP endpoint) → JSON-RPC error.
	out := rpcPost(t, handle.URL+"/mcp", "tools/list", map[string]any{})
	if _, ok := out["error"]; !ok {
		t.Fatalf("expected a JSON-RPC error for /mcp without an MCP profile; got %v", out)
	}
	if _, ok := out["result"]; ok {
		t.Fatalf("expected no result; got %v", out["result"])
	}
}

func TestMcpInboundTopLevelConfigBlock(t *testing.T) {
	tk, err := CreateToolkit(context.Background(), Options{
		McpConfig:  map[string]any{"mcpServer": map[string]any{"name": "from-config"}},
		Builtins:   false,
		ExtraTools: []Tool{echoTool},
	})
	if err != nil {
		t.Fatalf("CreateToolkit: %v", err)
	}
	defer tk.Close()

	// Serve() with no inline MCP falls back to the config block.
	handle, err := tk.Serve("127.0.0.1:0", ServeOptions{})
	if err != nil {
		t.Fatalf("Serve: %v", err)
	}
	defer handle.Stop()

	c, err := mcpclient.NewStreamableHttpClient(handle.URL + "/mcp")
	if err != nil {
		t.Fatalf("NewStreamableHttpClient: %v", err)
	}
	defer c.Close()
	if err := c.Start(context.Background()); err != nil {
		t.Fatalf("Start: %v", err)
	}
	res, err := c.Initialize(context.Background(), mcp.InitializeRequest{})
	if err != nil {
		t.Fatalf("Initialize: %v", err)
	}
	if res.ServerInfo.Name != "from-config" {
		t.Fatalf("serverInfo.name = %q, want from-config", res.ServerInfo.Name)
	}
}

// --- helpers -------------------------------------------------------------

func contains(ss []string, s string) bool {
	for _, x := range ss {
		if x == s {
			return true
		}
	}
	return false
}

// textOf extracts the first text content part of a CallToolResult.
func textOf(t *testing.T, res *mcp.CallToolResult) string {
	t.Helper()
	if len(res.Content) == 0 {
		t.Fatalf("result has no content")
	}
	tc, ok := mcp.AsTextContent(res.Content[0])
	if !ok {
		t.Fatalf("first content part is not text: %T", res.Content[0])
	}
	return tc.Text
}

// schemasEqual compares the served inputSchema against the original tool schema
// by normalizing both to generic JSON.
func schemasEqual(t *testing.T, got mcp.ToolInputSchema, want JSONSchema) bool {
	t.Helper()
	gb, err := json.Marshal(got)
	if err != nil {
		t.Fatalf("marshal got schema: %v", err)
	}
	wb, err := json.Marshal(want)
	if err != nil {
		t.Fatalf("marshal want schema: %v", err)
	}
	var gm, wm map[string]any
	_ = json.Unmarshal(gb, &gm)
	_ = json.Unmarshal(wb, &wm)
	return reflect.DeepEqual(gm, wm)
}
